azdev/operations/setup.py

```diff
@@ -258,14 +264,114 @@ def add_ext_repo(path):
         raise CLIError('Installation aborted.')
 
 
-def setup(cli_path=None, ext_repo_path=None, ext=None, deps=None):
+def _validate_input(cli_path=None, ext_repo_path=None, set_env=None, copy=None, use_global=None):
+    if copy and use_global:
+        raise CLIError("Copy and use global are mutally exlcusive.")
+    if not cli_path or not ext_repo_path:
+        if cli_path == "pypi" and any([use_global, copy, set_env]):
+            raise CLIError("pypi for cli path is mutally exlcusive with global copy and set env")
+        if (cli_path or ext_repo_path) and any([use_global, copy, set_env]):
+            raise CLIError("if global, copy, or set env are set then both an extensions repo "
+                           " and a cli repo must be specified")
+        if any([use_global, copy, set_env]):
+            raise CLIError("if global, copy, or set env are set then both an"
+                           " extensions repo and a cli repo must be specified")
+
 
-    require_virtual_env()
+def _check_paths(cli_path, ext_repo_path):
+    if not os.path.isdir(cli_path):
+        raise CLIError("The cli path is not a valid directory, please check the path")
+    if not os.path.isdir(ext_repo_path):
+        raise CLIError("The cli extensions path is not a valid directory, please check the path")
 
-    start = time.time()
+
+def setup(cli_path=None, ext_repo_path=None, ext=None, deps=None, set_env=None, copy=None, use_global=None):
+    if not set_env:
+        if not get_env_path():
+            raise CLIError('You are not running in a virtual enviroment and have not chosen to set one up.')
+    elif 'VIRTUAL_ENV' in os.environ:
+        raise CLIError("You are already running in a virtual enviroment, yet you want to set a new one up")
 
     heading('Azure CLI Dev Setup')
 
+    _validate_input(cli_path, ext_repo_path, set_env, copy, use_global)
+    # cases for handling legacy install
+    if not any([cli_path, ext_repo_path]) or cli_path == "pypi" or (not cli_path or not ext_repo_path):
+        return _handle_legacy(cli_path, ext_repo_path, ext, deps, time.time())
+    if 'CONDA_PREFIX' in os.environ:
+        raise CLIError('CONDA virutal enviroments are not supported outside'
+                       ' of interactive mode or when -c and -r are provided')
+    _check_paths(cli_path, ext_repo_path)
+
+    if set_env:
+        subprocess.call(shlex.split((const.VENV_CMD if const.IS_WINDOWS else const.VENV_CMD3) +
```

Comment:
Can we reuse the `py_cmd` function? `python -m venv {value of --set-env}` will do the same job.

Reply:
`py_cmd` calls the `cmd`, and `cmd` call `subprocess.check_output` who has implemented the error handling. Is it the same as `subprocess.check_call`? Even better it won't print the whole error stack to the screen.

BTW, all the steps involving in azdev of each are linear, and the virtual env created here is suppossed to be done such that we could do further operations among it. So, we should wait for it to be done and introducing unblock IO won't bring any benifict.

Reply:
The only difference between check_call and check_output is that check_ouput returns the stdout back to the caller and hides it from the user. In this case, we don't want the stdout returned to the caller, we want it displayed to the user immediately in case of failure. We also want to just crash here, not handle the subprocessor error, since this is just and unhandled failure. 

Not sure I get the second point here. Everything here in setup must be serial (linear) since all steps depend on their predecessor completing successfully.

Reply:
>  In this case, we don't want the stdout returned to the caller, we want it displayed to the user immediately in case of failure

I know this is a common strategy out there, many open source softwares adopt this way. 
However, among repos of Azure CLI, we would like to print a pretty result to user instead of the raw track stack.

As for the second point,  you mentioned 
> the IO should not be blocked from the user

So, I would guess you are thinking about leverage the async approach here.